lib/spack/spack/spec.py

```diff
@@ -1612,6 +1612,10 @@ def _concretize_helper(self, presets=None, visited=None):
         if self.name in visited:
             return False
 
+        if self.concrete and self.package.installed:
```

Comment:
I think it is sufficient and more correct in all of these cases to just check `self.concrete`.  If the spec is concrete, we shouldn't change it, even if it's not installed.  This should also be safe to do, since we only mark things concrete at the end of concretize now (since #5332 was merged)